File: algorithms/java/三步问题/三步问题.java
// Source : https://leetcode.cn/problems/three-steps-problem-lcci/
// Author : Yanan Wang
// Date   : 2023-06-12

class Solution {
    public int waysToStep(int n) {
        if (1 == n || 2 == n) return n;
        if (3 == n) return 4;
        final int MOD = 1000000007;
        int a = 1, b = 2, c = 4, d = 0;
        for (int i = 4; i <= n; i++) {
            d = (((a + b) % MOD) + c) % MOD;
            a = b;
            b = c;
            c = d;
        }
        return d;
    }
}

File: algorithms/python/三步问题/三步问题.py
# Source : https://leetcode.cn/problems/three-steps-problem-lcci/description/
# Author : Yanan Wang
# Date   : 2023-10-29

class Solution:
    def waysToStep(self, n: int) -> int:
        if n <= 2:
            return n
        a = b = 1
        c = 2
        for i in range(3, n + 1):
            ans = (a + b + c) % 1000000007
            a, b, c = b, c, ans
        return ans